File: src/ai/behaviorAnalyzer.ts
import { HistoryItem } from '../types';

export interface BehaviorInsights {
  repeatedRestaurants: Record<string, number>;
  repeatedFoods: Record<string, number>;
  dominantCuisine: string | null;
  messages: string[];
}

/**
 * Detects restaurant/food/cuisine repetition in recent history and
 * produces variety-encouraging (never health-shaming) messages.
 */
export function analyzeHistory(history: HistoryItem[]): BehaviorInsights {
  const restaurantCounts: Record<string, number> = {};
  const foodCounts: Record<string, number> = {};
  const cuisineCounts: Record<string, number> = {};

  history.forEach((h) => {
    restaurantCounts[h.restaurantName] = (restaurantCounts[h.restaurantName] || 0) + 1;
    foodCounts[h.foodName] = (foodCounts[h.foodName] || 0) + 1;
    cuisineCounts[h.cuisine] = (cuisineCounts[h.cuisine] || 0) + 1;
  });

  const messages: string[] = [];

  Object.entries(restaurantCounts).forEach(([name, count]) => {
    if (count >= 3) messages.push(`You've chosen ${name} ${count} times this week. Want to explore something different?`);
  });
  Object.entries(foodCounts).forEach(([name, count]) => {
    if (count >= 2) messages.push(`You've chosen ${name} several times recently. Here are some different options you may enjoy.`);
  });

  let dominantCuisine: string | null = null;
  const cuisineEntries = Object.entries(cuisineCounts).sort((a, b) => b[1] - a[1]);
  if (cuisineEntries.length) {
    dominantCuisine = cuisineEntries[0][0];
    if (history.length >= 4 && cuisineEntries[0][1] / history.length >= 0.6) {
      messages.push(`Most of your recent meals are ${dominantCuisine} cuisine. Try a highly rated different cuisine nearby.`);
    }
  }

  return {
    repeatedRestaurants: Object.fromEntries(Object.entries(restaurantCounts).filter(([, c]) => c >= 3)),
    repeatedFoods: Object.fromEntries(Object.entries(foodCounts).filter(([, c]) => c >= 2)),
    dominantCuisine,
    messages,
  };
}

File: backend/app/ai/behavior_analyzer.py
"""
Behavior-aware variety module.

Looks at a user's recent history rows and flags repetition at the
restaurant, food and cuisine level, exactly mirroring the frontend's
src/ai/behaviorAnalyzer.ts so both offline and connected modes agree.

Deliberately does NOT label anything "unhealthy" -- only surfaces
repetition and suggests variety, per the project's health-safety rules.
"""
from collections import Counter
from typing import List


def analyze_history(history_rows: List) -> dict:
    restaurant_counts = Counter(h.restaurant_name for h in history_rows)
    food_counts = Counter(h.food_name for h in history_rows)
    cuisine_counts = Counter(h.cuisine for h in history_rows)

    messages = []

    for name, count in restaurant_counts.items():
        if count >= 3:
            messages.append(f"You've chosen {name} {count} times recently. Want to explore something different?")

    for name, count in food_counts.items():
        if count >= 2:
            messages.append(f"You've chosen {name} several times recently. Here are some different options you may enjoy.")

    if cuisine_counts:
        top_cuisine, top_count = cuisine_counts.most_common(1)[0]
        if len(history_rows) >= 4 and top_count / len(history_rows) >= 0.6:
            messages.append(f"Most of your recent meals are {top_cuisine} cuisine. Try a highly rated different cuisine nearby.")

    return {
        "repeated_restaurants": {k: v for k, v in restaurant_counts.items() if v >= 3},
        "repeated_foods": {k: v for k, v in food_counts.items() if v >= 2},
        "dominant_cuisine": cuisine_counts.most_common(1)[0][0] if cuisine_counts else None,
        "messages": messages,
    }
